src/main/scala/scalaz/stream/io.scala

```diff
@@ -212,4 +217,173 @@ object io {
         else buf.take(m)
       }}
     }
+
+  /**
+   * Converts a source to a mutable `InputStream`.  The resulting input stream
+   * should be reasonably efficient and supports early termination (i.e. all
+   * finalizers associated with the input process will be run if the stream is
+   * closed).
+   *
+   * Takes a `copy` function which itself takes the following parameters in order:
+   *
+   * - `A` - the source chunk
+   * - `Int` - the logical offset into the source chunk
+   * - `Array[Byte]` - the target copy buffer
+   * - `Int` - the logical offset into the target buffer
+   * - `Int` - the length to copy
+   *
+   * The `copy` function returns the number of bytes copied, in total.  You should
+   * think of this function as a generic version of `System.arraycopy`, which is
+   * in fact the precise implementation you would use if `A = Array[Byte]`.  Note
+   * that `toInputStream` will never request a `copy` with a length value that
+   * exceeds the logical length of the source chunk as reported by the `logicalLength`
+   * function.  Additionally, `buffer` will always be *at least* `off2 + length` in
+   * size.  Thus, the `copy` function does not need to concern itself with bounds
+   * checking.
+   *
+   * The other function taken as a parameter is `logicalLength`, which simply returns
+   * the logical length of the source chunk in question.  Thus, if `A = Array[Byte]`,
+   * this function would simply be `{ _.length }`.
+   *
+   * @see #toInputStreamFromBytes
+   */
+  def toInputStream[A](p: Process[Task, A])(logicalLength: A => Int, copy: (A, Int, Array[Byte], Int, Int) => Int): InputStream = new InputStream {
```

Comment:
There is a convenience method which implements the type signature you suggest (element type of `Array[Byte]`).  The point of the more general signature is to avoid specializing on byte arrays, since not everyone uses them!  An obvious example of this is `ByteBuffer`, but there are other examples as well.  Forcing users into byte arrays requires extra copying in those use cases, which is often prohibitive.  The cost of invoking a buffering function of the signature used here is negligible compared to the cost of allocating, retaining and then discarding scratch space corresponding to every chunk.  The cost goes down even further (basically to zero, due to monomorphic call site inlining!) if we swap in a typeclass instead of a pair of functions.  This also makes the syntax a lot more palettable.

Per element computational costs arent that much of a concern, since one would expect that anyone caring about performance is going to be batching, and scalaz-stream already has plenty of element level costs.  What we want to avoid are *byte* level computational costs, and this is in fact precisely what we get around by allowing users to specify their own copy functions (note that functions like `arraycopy` do not have a byte level cost due to CPU intrinsics).